### PR TITLE
Implement Data.Reflectable

### LINKED ISSUE #286: Add Data.Reflectable
The [type-level integers](https://github.com/purescript/purescript/pull/4207) PR for the compiler was recently merged and it introduces automatic solving for the `Reflectable` class which is defined as the following.
```hs
module Data.Reflectable where

class Reflectable :: forall k. k -> Type -> Constraint
class Reflectable v t | v -> t where
  reflectType :: forall proxy. proxy v -> t
```
I'll open a PR once ES modules land in this repository.

### PR DESCRIPTION
**Pull request by purefunctor**
Closes #286. Adds a `Data.Reflectable` module which is where the compiler looks at for the automatically-solved `Reflectable` type class.

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [x] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Review by JordanMartinez — CHANGES_REQUESTED** [review]
Could you add tests for both the Reflectable and Reifiable type classes?

**Comment by purefunctor** [comment]
Should I copy-paste the tests from the compiler? https://github.com/purescript/purescript/blob/1d5c618054bc0c452128e8758dc01b758a3a6f39/tests/purs/passing/SolvingReflectable.purs

I think testing `Reflectable` here would be a bit redundant given the compiler tests. Similarly, tests for `IsSymbol` is done there too. I'm fine with either way, though :sweat_smile:

File: src/Data/Reflectable.purs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+instance Reifiable String
+
+-- local definition for use in `reifyType`
+foreign import unsafeCoerce :: forall a b. a -> b
+
+-- | Reify a value of type `t` such that it can be consumed by a
+-- | function constrained by the `Reflectable` type class. For
+-- | example:
+-- |
+-- | ```hs
```
Comment (thomashoneyman) [inline src/Data/Reflectable.purs]:
```suggestion
-- | ```purs
```

**Comment by JordanMartinez** [comment]
> Should I copy-paste the tests from the compiler? https://github.com/purescript/purescript/blob/1d5c618054bc0c452128e8758dc01b758a3a6f39/tests/purs/passing/SolvingReflectable.purs
> 
> I think testing `Reflectable` here would be a bit redundant given the compiler tests. Similarly, tests for `IsSymbol` is done there too. I'm fine with either way, though :sweat_smile:

I think adding it here forces other backends to also implement the tests, which aren't being tested in the compiler. We'd also probably want something more granular.
```purs
testReflection :: AlmostEff
testReflection = do
  assert "Symbol reflection" $ reflectType (Proxy :: Proxy "foo") == "foo"
  assert "Int reflection - 1" $ reflectType (Proxy :: Proxy 1) == 1
  assert "Int reflection - 0" $ reflectType (Proxy :: Proxy 0) == 0
  assert "Int reflection - (-1)" $ reflectType (Proxy :: Proxy (-1)) == (-1)
  assert "Boolean - True reflection" $ reflectType (Proxy :: Proxy PB.True) == true
  assert "Boolean - False reflection" $ reflectType (Proxy :: Proxy PB.False) == false
  assert "Ordering - LT reflection" $ reflectType (Proxy :: Proxy PO.LT) == LT
  assert "Ordering - EQ reflection" $ reflectType (Proxy :: Proxy PO.EQ) == EQ
  assert "Ordering - GT reflection" $ reflectType (Proxy :: Proxy PO.GT) == GT
```
Similarly for the `Reifiable` which isn't being tested by anything, right?

**Review by JordanMartinez — APPROVED** [review]

**Comment by JordanMartinez** [comment]
:ping_pong: @thomashoneyman

**Review by thomashoneyman — APPROVED** [review]
👍 thanks, @PureFunctor!

**Pull request merged**